src/XrdTpc/XrdTpcNSSSupport.cc

```diff
@@ -0,0 +1,262 @@
+     * is found, then it is written into the current
+     * tempfile.
+     *
+     * The fname argument is used solely for debugging.
+     * 
+     * Returns true on success.
+     */
+    bool processFile(file_smart_ptr &fd, const std::string &fname);
+
+private:
+    XrdSysError &m_log;
+
+        // Grid CA directories tend to keep everything in triplicate;
+        // we keep a unique hash of all known CAs so we write out each
+        // one only once.
+    std::unordered_set<std::string> m_known_cas;
+    const int m_output_fd;
+
+    decltype(&XrdCryptosslX509ToFile) m_x509_to_file_func;
+    int (*m_file_to_x509_func)(FILE *file, XrdCryptoX509Chain *c);
+};
+
+
+bool
+CASet::processFile(file_smart_ptr &fp, const std::string &fname)
+{
+    XrdCryptoX509Chain chain;
+    // Not checking return value here; function returns `0` on error and
+    // if no certificate is found.
+    (*m_file_to_x509_func)(fp.get(), &chain);
+
+    auto ca = chain.Begin();
+    // Note we purposely leak the outputfp here; we are just borrowing the handle.
+    FILE *outputfp = fdopen(m_output_fd, "w");
+    if (!outputfp) {
+        m_log.Emsg("CAset", "Failed to reopen file for output", fname.c_str());
+        return false;
+    }
+    while (ca) {
+        auto hash_ptr = ca->SubjectHash();
+        if (!hash_ptr) {
+            continue;
+        }
+        auto iter = m_known_cas.find(hash_ptr);
+        if (iter != m_known_cas.end()) {
+            //m_log.Emsg("CAset", "Skipping known CA with hash", fname.c_str(), hash_ptr);
+            ca = chain.Next();
+            continue;
+        }
+        //m_log.Emsg("CAset", "New CA with hash", fname.c_str(), hash_ptr);
+        m_known_cas.insert(hash_ptr);
+
+        if ((*m_x509_to_file_func)(ca, outputfp)) {
+            m_log.Emsg("CAset", "Failed to write out CA", fname.c_str());
+            return false;
+        }
+        ca = chain.Next();
+    }
+    fflush(outputfp);
+
+    return true;
+}
+
+}
+
+
+using namespace TPC;
+
+
+std::unique_ptr<XrdTpcNSSSupport::TempCAGuard>
+XrdTpcNSSSupport::TempCAGuard::create(XrdSysError &err) {
+    char fname[] = "/tmp/xrootd_curl_nss_ca.XXXXXX.pem";
+    int fd = mkstemps(fname, 4);
+    if (fd < 0) {
+        err.Emsg("NSSSupport", "Failed to create temp file:", strerror(errno));
+        return std::unique_ptr<TempCAGuard>();
+    }
+    return std::unique_ptr<TempCAGuard>(new TempCAGuard(fd, fname));
+}
+
+
+XrdTpcNSSSupport::TempCAGuard::~TempCAGuard() {
+    if (m_fd >= 0) {
+        unlink(m_fname.c_str());
+        close(m_fd);
+    }
+}
+
+
+XrdTpcNSSSupport::TempCAGuard::TempCAGuard(int fd, const std::string &fname)
+    : m_fd(fd), m_fname(fname)
+    {}
+
+
+XrdTpcNSSSupport::XrdTpcNSSSupport(XrdSysError *err, std::string ca_dir)
+    : m_log(*err),
+      m_ca_dir(ca_dir),
+      m_plugin(err, &g_ver, "Crypto Support", "libXrdCryptossl.so")
+{
+    if (!(m_x509_to_file_func = m_plugin.Resolve("XrdCryptosslX509ToFile")) ||
```

Comment:
Why in the world are you manually resolving these? The functions are available in standard shared libraries that you can link with. Presumably, this code is so critical to prevent HttpTpc from blowing up that you would really want it to work instead of just skipping all of this if you can't find the methods. Plus it makes no sense anyway since the libraries where you added these methods will be distributed with the code that uses them. So, why do this?

Reply:
Actually, it appears we cannot link to this - that's why I had to go this direction (trust me, already tried!  It's a pain to load like this and muck around with function pointers).  CMake cowardly refuses to link to a shared module (as opposed to a shared library).

Another option would be to compile + link an extra copy of those functions just for HttpTpc; I don't particularly like that either.

I like the idea of doing a hard fail instead of falling back.  Will work on that.

Reply:
We now do a hard fail when the initialization fails.

Resolving this conversation because the discussion about how to best include the `XrdCryptosslAux` functions continued on below.